File: physics/fluid_sim.js
// Define the simulation parameters
const nx = 101;      // Grid size
const ny = 101;
const nt = 100;      // Number of time steps
const dx = 1;        // Grid spacing
const dt = 1;        // Time step
const viscosity = 0.1; // Fluid viscosity

// Initialize the fluid velocity and density
let u = Array(ny).fill().map(() => Array(nx).fill(0));
let v = Array(ny).fill().map(() => Array(nx).fill(0));
let rho = Array(ny).fill().map(() => Array(nx).fill(1));

// Define the LBM constants
const omega = 1.9;
const tau = viscosity * (3 * omega - 0.5) / (dx ** 2);

// Define the LBM weights and direction vectors
const w = [4/9, 1/9, 1/9, 1/9, 1/9, 1/36, 1/36, 1/36, 1/36];
const cx = [0, 1, 0, -1, 0, 1, -1, -1, 1];
const cy = [0, 0, 1, 0, -1, 1, 1, -1, -1];

// Define a function to calculate the equilibrium distribution
function equilibrium(rho, u, v) {
    const cu = cx.map((x, i) => 3 * (x * u + cy[i] * v));
    const uu = 3/2 * (u ** 2 + v ** 2);
    const feq = Array.from(Array(9), () => Array(ny).fill().map(() => Array(nx).fill(0)));
    for (let i = 0; i < 9; i++) {
        feq[i] = rho.map(row => row.slice()).map((row, j) => row.map((val, k) => {
            return rho[j][k] * w[i] * (1 + cu[i][j][k] + 0.5 * cu[i][j][k] ** 2 - uu[j][k]);
        }));
    }
    return feq;
}

// Define a function to calculate the macroscopic variables
function macroscopic(f) {
    const rho = f.reduce((acc, val) => acc.map((row, i) => row.map((col, j) => col + val[i][j])), Array(ny).fill().map(() => Array(nx).fill(0)));
    const u = f.map((val, i) => rho.map(row => row.slice()).map((row, j) => row.map((val2, k) => val[i][j][k] * cx[i]))).reduce((acc, val) => acc.map((row, i) => row.map((col, j) => col + val[i][j])), Array(ny).fill().map(() => Array(nx).fill(0))) / rho;
    const v = f.map((val, i) => rho.map(row => row.slice()).map((row, j) => row.map((val2, k) => val[i][j][k] * cy[i]))).reduce((acc, val) => acc.map((row, i) => row.map((col, j) => col + val[i][j])), Array(ny).fill().map(() => Array(nx).fill(0))) / rho;
    return [rho, u, v];
}

// Define the main simulation loop
for (let n = 0; n < nt; n++) {
	// Calculate the equilibrium distribution
	const feq = equilibrium(rho, u, v);
	
	// Calculate the LBM collision step
	const fcoll = feq.map((val, i) => val.map((val2, j) => val2.map((val3, k) => (1 - omega) * val3 + omega * f[i][j][k])));
	
	// Calculate the LBM streaming step
	const fstream = Array.from(Array(9), () => Array(ny).fill().map(() => Array(nx).fill(0)));
	for (let i = 0; i < 9; i++) {
		const xi = cx[i];
		const yi = cy[i];
		fstream[i] = fcoll.map((row, j) => row.map((val, k) => {
			const x = k - xi;
			const y = j - yi;
			if (x >= nx) x -= nx;
			if (x < 0) x += nx;
			if (y >= ny) y -= ny;
			if (y < 0) y += ny;
			return fcoll[y][x][k];
		}));
	}

	// Update the macroscopic variables
	[rho, u, v] = macroscopic(fstream);

	// Apply the boundary conditions
	u[0][1] = 0.1;
	u[0][nx - 2] = 0.1;
	u[ny - 1][1] = -0.1;
	u[ny - 1][nx - 2] = -0.1;
	v[0][1] = 0;
	v[0][nx - 2] = 0;
	v[ny - 1][1] = 0;
	v[ny - 1][nx - 2] = 0;

	// Update the fluid density
	rho = rho.map((row, i) => row.map((val, j) => val + dt * (u[i][j] + v[i][j])));

	// Update the fluid velocity
	const force = Array(ny).fill().map(() => Array(nx).fill(0));
	for (let i = 1; i < ny - 1; i++) {
		for (let j = 1; j < nx - 1; j++) {
			force[i][j] = -1 * viscosity * (u[i + 1][j] - 2 * u[i][j] + u[i - 1][j]) / (dx ** 2) - rho[i][j] * 9.8;
		}
	}
	u = u.map((row, i) => row.map((val, j) => val + dt * force[i][j]));
	v = v.map((row, i) => row.map((val, j) => val + dt * force[i][j]));

	// Display the current frame
	console.clear();
	console.log(`Time step: ${n + 1} / ${nt}`);
	console.table(rho);
}

/*

Note that this script is designed to be run in the browser console. It sets up a 2D grid with 101 rows and 101 columns, and then simulates fluid dynamics for 100 time steps. The script calculates the equilibrium distribution, performs the LBM collision and streaming steps, updates the macroscopic variables, applies boundary conditions, updates the fluid density, and updates the fluid velocity
*/

File: physics/fluid_sim.py
#!/usr/bin/python3

import numpy as np
import matplotlib.pyplot as plt

# Define the simulation parameters
nx, ny = 101, 101  # Grid size
nt = 100           # Number of time steps
dx = 1             # Grid spacing
dt = 1             # Time step
viscosity = 0.1    # Fluid viscosity

# Initialize the fluid velocity and density
u = np.zeros((ny, nx))
v = np.zeros((ny, nx))
rho = np.ones((ny, nx))

# Define the LBM constants
omega = 1.9
tau = viscosity * (3 * omega - 0.5) / (dx ** 2)

# Define the LBM weights and direction vectors
w = np.array([4/9, 1/9, 1/9, 1/9, 1/9, 1/36, 1/36, 1/36, 1/36])
cx = np.array([0, 1, 0, -1, 0, 1, -1, -1, 1])
cy = np.array([0, 0, 1, 0, -1, 1, 1, -1, -1])

# Define a function to calculate the equilibrium distribution
def equilibrium(rho, u, v):
    cu = 3 * (cx * u + cy * v)
    uu = 3/2 * (u ** 2 + v ** 2)
    feq = np.zeros((9, ny, nx))
    for i in range(9):
        feq[i] = rho * w[i] * (1 + cu[i] + 0.5 * cu[i] ** 2 - uu)
    return feq

# Define a function to calculate the macroscopic variables
def macroscopic(f):
    rho = np.sum(f, axis=0)
    u = np.sum(f * cx, axis=0) / rho
    v = np.sum(f * cy, axis=0) / rho
    return rho, u, v

# Main loop
for n in range(nt):
    # Calculate the equilibrium distribution
    feq = equilibrium(rho, u, v)
    # Calculate the non-equilibrium distribution
    fneq = np.zeros((9, ny, nx))
    for i in range(9):
        cu = 3 * (cx[i] * u + cy[i] * v)
        fneq[i] = w[i] * rho * cu / (3 * (tau + 0.5))
    # Update the distribution function
    f = feq + fneq - (feq - equilibrium(rho, u, v)) / tau
    # Calculate the macroscopic variables
    rho, u, v = macroscopic(f)
    # Apply boundary conditions (bounce-back)
    u[0, :] = 0
    v[0, :] = 0
    u[-1, :] = 0
    v[-1, :] = 0
    u[:, 0] = 0
    v[:, 0] = 0
    u[:, -1] = 0
    v[:, -1] = 0
    # Plot the results
    if n % 10 == 0:
        plt.clf()
        plt.quiver(u, v)
        plt.pause(0.001)

plt.show()
